# Using ROBOKOP's Quick Question Answering Service

The expand, enrich, and similarity services each offer bite-sized approaches to interacting with ROBOKOP.  The ROBOKOP "quick" service offers a slightly more complex approach to answering questions.   We'll be using the following function to call the quick service.  Note that calling the service requires posting a question.  We'll discuss the format of the question below.

In [1]:
robokop_server='robokop.renci.org'

In [2]:
import requests

def quick(question,max_results=None):
    url=f'http://{robokop_server}:80/api/simple/quick/'
    if max_results is not None:
        url += f'?max_results={max_results}'
    response = requests.post(url,json=question)
    print( f"Return Status: {response.status_code}" )
    if response.status_code == 200:
        return response.json()
    return response

## Question Format and Basic Usage

The question is a python dictionary.  It takes a key `machine_question`, which is a dictionary containing a list of `nodes` and a list of `edges`.   Each node object needs an integer `id`.   Any node can also have a `curie` specifying that node.  The edges define connections between the nodes using the identifiers of the nodes as source and targets. The following function shows how to construct a single-hop question.  It takes a specified node of `type1` and looks for any node of `type2`.

In [3]:
def make_one_step_question(type1, id1, type2,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [
                        {
                            'id': 0,
                            'curie': id1,
                            'type': type1
                        },
                        {
                            'id': 1,
                            'type': type2
                        }
                    ],
                    'edges': [
                        {
                            'source_id': 0,
                            'target_id': 1
                        }
                    ]
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    return question

Here, we will specify a one-hop question asking for the phenotypes associated with Fanconi Anemia (MONDO:0019391).  We first construct the question then use it to call the quick service.  In fact, this is how the expand function is implemented internally.  The output is returned in the KG standard answer format.

In [4]:
q = make_one_step_question('disease','MONDO:0019391','phenotypic_feature')
r = quick(q)

Return Status: 200


In [5]:
# Uncomment to see a big long answer
#import json
#print( json.dumps(r, indent=4))

The output can be simplified with a function like this:

In [6]:
import pandas as pd

def parse_answer(returnanswer):
    nodes = [answer['nodes'][1] for answer in returnanswer['answers']]
    edges = [answer['edges'][0] for answer in returnanswer['answers']]
    answers = [ {"result_id": node["id"], 
                 "result_name": node["name"], 
                 "type": edge["type"],
                 "source": edge['edge_source']}
              for node,edge in zip(nodes,edges)]
    return pd.DataFrame(answers)

In [7]:
parse_answer(r)

result_id                                        result_name  \
0    HP:0005528                        Bone marrow hypocellularity   
1    HP:0004810                      Congenital hypoplastic anemia   
2    HP:0001908                                 Hypoplastic anemia   
3    HP:0010972                    Anemia of inadequate production   
4    HP:0004820                      Acute myelomonocytic leukemia   
5    HP:0000953                      Hyperpigmentation of the skin   
6    HP:0001972                                  Macrocytic anemia   
7    HP:0001876                                       Pancytopenia   
8    HP:0003974                                      Absent radius   
9    HP:0001905                        Congenital thrombocytopenia   
10   HP:0000286                                         Epicanthus   
11   HP:0003254                          Abnormality of DNA repair   
12   HP:0000957                                  Cafe-au-lait spot   
13   HP:0004859                   Amegakaryocytic thrombocytopenia   
14   HP:0003468                      Abnormal vertebral morphology   
15   HP:0002863                                     Myelodysplasia   
16   HP:0002023                                       Anal atresia   
17   HP:0009777                                       Absent thumb   
18   HP:0002984                           Hypoplasia of the radius   
19   HP:0004430                   Severe combined immunodeficiency   
20   HP:0005510                       Transient erythroblastopenia   
21   HP:0001172                           Abnormality of the thumb   
22   HP:0000824                          Growth hormone deficiency   
23   HP:0005505                                  Refractory anemia   
24   HP:0001000                   Abnormality of skin pigmentation   
25   HP:0001896                                  Reticulocytopenia   
26   HP:0004808                             Acute myeloid leukemia   
27   HP:0000104                                     Renal agenesis   
28   HP:0000164                       Abnormality of the dentition   
29   HP:0000086                                     Ectopic kidney   
..          ...                                                ...   
156  HP:0005280                             Depressed nasal bridge   
157  HP:0005522         Pyridoxine-responsive sideroblastic anemia   
158  HP:0009943              Complete duplication of thumb phalanx   
159  HP:0000125                                      Pelvic kidney   
160  HP:0008678                           Renal hypoplasia/aplasia   
161  HP:0003213  Deficient excision of UV-induced pyrimidine di...   
162  HP:0007400                        Irregular hyperpigmentation   
163  HP:0000453                                    Choanal atresia   
164  HP:0001537                                   Umbilical hernia   
165  HP:0001284                                          Areflexia   
166  HP:0000072                                        Hydroureter   
167  HP:0000010                 Recurrent urinary tract infections   
168  HP:0003774                     Stage 5 chronic kidney disease   
169  HP:0000083                                Renal insufficiency   
170  HP:0004349                       Reduced bone mineral density   
171  HP:0001562                                    Oligohydramnios   
172  HP:0006824                            Cranial nerve paralysis   
173  HP:0007354                      Amyotrophic lateral sclerosis   
174  HP:0002245                                Meckel diverticulum   
175  HP:0001762                                Talipes equinovarus   
176  HP:0000027                                        Azoospermia   
177  HP:0001824                                        Weight loss   
178  HP:0002414                                       Spina bifida   
179  HP:0002251                              Aganglionic megacolon   
180  HP:0000483                                        Astigmatism   
181  HP:0000505 

## Multi-step questions

It's straightforward to generalize this linear query to an N-item path.  This function constructs a question from a list of types and a list of identifiers.  The types are the types of nodes that will be traversed along the path, and the identifiers represent fixed elements in the path.  The length of types and ids should be equal, and free nodes should specify an id of None:

In [8]:
def make_N_step_question(types,curies,rebuild = None):
    question = {
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild'] = 'True'
    for i,t in enumerate(types):
        newnode = {'id': i, 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        if i > 0:
            question['machine_question']['edges'].append( {'source_id': i-1, 'target_id': i})
    return question

We can recapitulate our previous question with this new function like this:

In [9]:
newq = make_N_step_question(['disease','phenotypic_feature'],['MONDO:0019391',None])
q == newq

True

Now we could expand to a longer query.  The following question will start at the disease MONDO:0019391, go to a gene, and from there to a biological process or activity.

In [10]:
two_step_question = make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,None])

In [11]:
two_step_answer = quick(two_step_question)

Return Status: 200


We can extract the node names along the paths with a simple function:

In [12]:
def extract_node_names(returnanswer):
    nodes = [{f'node_{i}': (node['name'] if 'name' in node else node['id'])  
              for i,node in enumerate(answer['nodes'])} for answer in returnanswer['answers']]
    return pd.DataFrame(nodes)

In [13]:
two_step_nodes = extract_node_names(two_step_answer)
two_step_nodes

node_0   node_1  \
0    Fanconi anemia   FANCD2   
1    Fanconi anemia   FANCD2   
2    Fanconi anemia   FANCD2   
3    Fanconi anemia   FANCD2   
4    Fanconi anemia   FANCD2   
5    Fanconi anemia   FANCD2   
6    Fanconi anemia    FANCM   
7    Fanconi anemia    FANCM   
8    Fanconi anemia    FANCA   
9    Fanconi anemia    FANCA   
10   Fanconi anemia    FANCA   
11   Fanconi anemia    FANCA   
12   Fanconi anemia    FANCA   
13   Fanconi anemia    FANCA   
14   Fanconi anemia    FANCG   
15   Fanconi anemia    FANCG   
16   Fanconi anemia    FANCI   
17   Fanconi anemia    FANCI   
18   Fanconi anemia    FANCI   
19   Fanconi anemia    FANCI   
20   Fanconi anemia    FANCM   
21   Fanconi anemia    FANCM   
22   Fanconi anemia    FANCM   
23   Fanconi anemia    FANCM   
24   Fanconi anemia    FANCC   
25   Fanconi anemia    FANCC   
26   Fanconi anemia    FANCC   
27   Fanconi anemia    FANCC   
28   Fanconi anemia     SLX4   
29   Fanconi anemia     SLX4   
..              ...      ...   
220  Fanconi anemia   RAD51D   
221  Fanconi anemia   RAD51D   
222  Fanconi anemia     FAN1   
223  Fanconi anemia     FAN1   
224  Fanconi anemia    BRCA2   
225  Fanconi anemia    BRCA2   
226  Fanconi anemia    BRCA2   
227  Fanconi anemia    RAD51   
228  Fanconi anemia    RAD51   
229  Fanconi anemia    RAD51   
230  Fanconi anemia    RAD51   
231  Fanconi anemia    RAD51   
232  Fanconi anemia    RAD51   
233  Fanconi anemia   RAD51D   
234  Fanconi anemia   RAD51D   
235  Fanconi anemia  TP53BP1   
236  Fanconi anemia  TP53BP1   
237  Fanconi anemia  TP53BP1   
238  Fanconi anemia  TP53BP1   
239  Fanconi anemia    FANCC   
240  Fanconi anemia    FANCC   
241  Fanconi anemia    FANCC   
242  Fanconi anemia    FANCC   
243  Fanconi anemia     DNA2   
244  Fanconi anemia     DNA2   
245  Fanconi anemia     DNA2   
246  Fanconi anemia     DNA2   
247  Fanconi anemia     RNF8   
248  Fanconi anemia     RNF8   
249  Fanconi anemia     RNF8   

                                                node_2  
0                        interstrand cross-link repair  
1                        interstrand cross-link repair  
2                        interstrand cross-link repair  
3                        interstrand cross-link repair  
4                        interstrand cross-link repair  
5                        interstrand cross-link repair  
6                        interstrand cross-link repair  
7                        interstrand cross-link repair  
8                        interstrand cross-link repair  
9                        interstrand cross-link repair  
10                       interstrand cross-link repair  
11                       interstrand cross-link repair  
12                       interstrand cross-link repair  
13                       interstrand cross-link repair  
14                       interstrand cross-link repair  
15                       interstrand cross-link repair  
16                       interstrand cross-link repair  
17                       interstrand cross-link repair  
18                       interstrand cross-link repair  
19                       interstrand cross-link repair  
20                       interstrand cross-link repair  
21                       interstrand cross-link repair  
22                       interstrand cross-link repair  
23                       interstrand cross-link repair  
24                       interstrand cross-link repair  
25                       interstrand cross-link repair  
26                       interstrand cross-link repair  
27                       interstrand cross-link repair  
28                       interstrand cross-link repair  
29                       interstrand cross-link repair  
..                                                 ...  
220                               telomere maintenance  
221                               telomere maintenance  
222           cellular response to DNA damage stimulus  
223           cellular response t

At the node name level there are what appear to be several repeated paths.  In fact, these paths differ in the predicates, which we are not showing:

In [14]:
print(f"The first answer's second edge comes from {two_step_answer['answers'][0]['edges'][1]['edge_source']}")
print(f"The second answer's second edge comes from {two_step_answer['answers'][1]['edges'][1]['edge_source']}")

The first answer's second edge comes from quickgo.go_term_to_gene_annotation
The second answer's second edge comes from biolink.gene_get_process_or_function


## Specifying multiple fixed nodes

The examples above have all started from one known node and expanded from it, sometimes in multiple steps.  It's also possible to have more than one node specified.  For instance, if we wanted to look for genes that link Fanconi Aneima (MONDO:0019391) and DNA repair (GO:0006281), we could rerun our previous query, but setting the final curie, like this:

In [15]:
two_step_question_fixed_ends = \
   make_N_step_question(['disease','gene','biological_process_or_activity'],['MONDO:0019391',None,'GO:0006281'])
two_step_answer_fixed_ends = quick(two_step_question_fixed_ends)

Return Status: 200


In [16]:
extract_node_names(two_step_answer_fixed_ends)

node_0    node_1      node_2
0   Fanconi anemia    FANCD2  DNA repair
1   Fanconi anemia    FANCD2  DNA repair
2   Fanconi anemia    FANCD2  DNA repair
3   Fanconi anemia    FANCD2  DNA repair
4   Fanconi anemia    FANCD2  DNA repair
5   Fanconi anemia    FANCD2  DNA repair
6   Fanconi anemia     FANCA  DNA repair
7   Fanconi anemia     FANCA  DNA repair
8   Fanconi anemia     FANCA  DNA repair
9   Fanconi anemia     FANCA  DNA repair
10  Fanconi anemia     FANCA  DNA repair
11  Fanconi anemia     FANCA  DNA repair
12  Fanconi anemia       XPA  DNA repair
13  Fanconi anemia       XPA  DNA repair
14  Fanconi anemia     FANCG  DNA repair
15  Fanconi anemia     FANCG  DNA repair
16  Fanconi anemia     FANCC  DNA repair
17  Fanconi anemia     FANCC  DNA repair
18  Fanconi anemia     FANCC  DNA repair
19  Fanconi anemia     FANCC  DNA repair
20  Fanconi anemia     RAD51  DNA repair
21  Fanconi anemia     RAD51  DNA repair
22  Fanconi anemia     RAD51  DNA repair
23  Fanconi anemia     RAD51  DNA repair
24  Fanconi anemia     RAD51  DNA repair
25  Fanconi anemia     RAD51  DNA repair
26  Fanconi anemia    RAD51C  DNA repair
27  Fanconi anemia    RAD51C  DNA repair
28  Fanconi anemia       WRN  DNA repair
29  Fanconi anemia       WRN  DNA repair
..             ...       ...         ...
66  Fanconi anemia    RAD51D  DNA repair
67  Fanconi anemia    RAD51D  DNA repair
68  Fanconi anemia    RAD51D  DNA repair
69  Fanconi anemia    RAD51D  DNA repair
70  Fanconi anemia     BRCA2  DNA repair
71  Fanconi anemia     BRCA2  DNA repair
72  Fanconi anemia     BRCA2  DNA repair
73  Fanconi anemia     UBE2T  DNA repair
74  Fanconi anemia     UBE2T  DNA repair
75  Fanconi anemia     UBE2T  DNA repair
76  Fanconi anemia  RAD51AP1  DNA repair
77  Fanconi anemia  RAD51AP1  DNA repair
78  Fanconi anemia  RAD51AP1  DNA repair
79  Fanconi anemia  RAD51AP1  DNA repair
80  Fanconi anemia       ATM  DNA repair
81  Fanconi anemia       ATM  DNA repair
82  Fanconi anemia     DDX11  DNA repair
83  Fanconi anemia     DDX11  DNA repair
84  Fanconi anemia      EME1  DNA repair
85  Fanconi anemia      EME1  DNA repair
86  Fanconi anemia     SLX1B  DNA repair
87  Fanconi anemia     SLX1B  DNA repair
88  Fanconi anemia     SLX1A  DNA repair
89  Fanconi anemia     SLX1A  DNA repair
90  Fanconi anemia     CENPX  DNA repair
91  Fanconi anemia     CENPX  DNA repair
92  Fanconi anemia     CENPS  DNA repair
93  Fanconi anemia     CENPS  DNA repair
94  Fanconi anemia    BOD1L1  DNA repair
95  Fanconi anemia    BOD1L1  DNA repair

[96 rows x 3 columns]

## Non-path queries

So far, we've only looked at linear paths.  But the question format is actually more general than that - we can define a path pattern generally. So for instance, in the above query, we find all genes that are linked to both FA and DNA repair.  But what if we wanted to find entities that are connected to more than two specified entities.  Here is a query-generation function for the star query:

In [17]:
def make_star_question(types,curies,shared_type,rebuild=None):
    """Create a question to find entities of shared_type that are linked to all of the nodes specified in the
    types and curies arrays."""
    question = {
                #'rebuild': 'True',
                'machine_question': {
                    'nodes': [],
                    'edges': []
                }
            }
    if rebuild is not None and str(rebuild).upper() == 'TRUE':
        question['rebuild']='True'
    question['machine_question']['nodes'].append( {'id': 0, 'type': shared_type})
    for i,t in enumerate(types):
        newnode = {'id': i+1, 'type': t}
        if curies[i] is not None:
            newnode['curie'] = curies[i]
        question['machine_question']['nodes'].append(newnode)
        question['machine_question']['edges'].append( {'source_id': 0, 'target_id': i+1})
    return question

Suppose I have this set of GO terms, and I'd like to find genes that they all have in common:

* 'voltage-gated sodium channel activity(GO:0005248)',
* 'muscle contraction(GO:0006936)',
* 'voltage-gated ion channel activity(GO:0005244)',
* 'regulation of ion transmembrane transport(GO:0034765)',
* 'sodium ion transmembrane transport(GO:0035725)',
* 'neuronal action potential(GO:0019228)',
* 'membrane depolarization during action potential(GO:0086010)',
* 'sodium ion transport(GO:0006814)'

In [18]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814']
types = ['biological_process_or_activity' for g in go_terms]
star_q = make_star_question(types,go_terms,'gene')
star_q

{'machine_question': {'edges': [{'source_id': 0, 'target_id': 1},
   {'source_id': 0, 'target_id': 2},
   {'source_id': 0, 'target_id': 3},
   {'source_id': 0, 'target_id': 4},
   {'source_id': 0, 'target_id': 5},
   {'source_id': 0, 'target_id': 6},
   {'source_id': 0, 'target_id': 7},
   {'source_id': 0, 'target_id': 8}],
  'nodes': [{'id': 0, 'type': 'gene'},
   {'curie': 'GO:0005248', 'id': 1, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936', 'id': 2, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244', 'id': 3, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765', 'id': 4, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725', 'id': 5, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0019228', 'id': 6, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0086010', 'id': 7, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006814',
    'id': 8,
    'type': 'biological_process_or_activit

In [19]:
common_gene_answer = quick(star_q)

Return Status: 200


In [20]:
extract_node_names(common_gene_answer)

node_0                                 node_1              node_2  \
0    SCN4A  voltage-gated sodium channel activity  muscle contraction   
1    SCN4A  voltage-gated sodium channel activity  muscle contraction   
2    SCN4A  voltage-gated sodium channel activity  muscle contraction   
3    SCN4A  voltage-gated sodium channel activity  muscle contraction   
4    SCN4A  voltage-gated sodium channel activity  muscle contraction   
5    SCN4A  voltage-gated sodium channel activity  muscle contraction   
6    SCN4A  voltage-gated sodium channel activity  muscle contraction   
7    SCN4A  voltage-gated sodium channel activity  muscle contraction   
8    SCN4A  voltage-gated sodium channel activity  muscle contraction   
9    SCN4A  voltage-gated sodium channel activity  muscle contraction   
10   SCN4A  voltage-gated sodium channel activity  muscle contraction   
11   SCN4A  voltage-gated sodium channel activity  muscle contraction   
12   SCN4A  voltage-gated sodium channel activity  muscle contraction   
13   SCN4A  voltage-gated sodium channel activity  muscle contraction   
14   SCN4A  voltage-gated sodium channel activity  muscle contraction   
15   SCN4A  voltage-gated sodium channel activity  muscle contraction   
16   SCN4A  voltage-gated sodium channel activity  muscle contraction   
17   SCN4A  voltage-gated sodium channel activity  muscle contraction   
18   SCN4A  voltage-gated sodium channel activity  muscle contraction   
19   SCN4A  voltage-gated sodium channel activity  muscle contraction   
20   SCN4A  voltage-gated sodium channel activity  muscle contraction   
21   SCN4A  voltage-gated sodium channel activity  muscle contraction   
22   SCN4A  voltage-gated sodium channel activity  muscle contraction   
23   SCN4A  voltage-gated sodium channel activity  muscle contraction   
24   SCN4A  voltage-gated sodium channel activity  muscle contraction   
25   SCN4A  voltage-gated sodium channel activity  muscle contraction   
26   SCN4A  voltage-gated sodium channel activity  muscle contraction   
27   SCN4A  voltage-gated sodium channel activity  muscle contraction   
28   SCN4A  voltage-gated sodium channel activity  muscle contraction   
29   SCN4A  voltage-gated sodium channel activity  muscle contraction   
..     ...                                    ...                 ...   
220  SCN4A  voltage-gated sodium channel activity  muscle contraction   
221  SCN4A  voltage-gated sodium channel activity  muscle contraction   
222  SCN4A  voltage-gated sodium channel activity  muscle contraction   
223  SCN4A  voltage-gated sodium channel activity  muscle contraction   
224  SCN4A  voltage-gated sodium channel activity  muscle contraction   
225  SCN4A  voltage-gated sodium channel activity  muscle contraction   
226  SCN4A  voltage-gated sodium channel activity  muscle contraction   
227  SCN4A  voltage-gated sodium channel activity  muscle contraction   
228  SCN4A  voltage-gated sodium channel activity  muscle contraction   
229  SCN4A  voltage-gated sodium channel activity  muscle contraction   
230  SCN4A  voltage-gated sodium channel activity  muscle contraction   
231  SCN4A  voltage-gated sodium channel activity  muscle contraction   
232  SCN4A  voltage-gated sodium channel activity  muscle contraction   
233  SCN4A  voltage-gated sodium channel activity  muscle contraction   
234  SCN4A  voltage-gated sodium channel activity  muscle contraction   
235  SCN4A  voltage-gated sodium channel activity  muscle contraction   
236  SCN4A  voltage-gated sodium channel activity  muscle contraction   
237  SCN4A  voltage-gated sodium channel activity  muscle contraction   
238  SCN4A  voltage-gated sodium channel activity  muscle contraction   
239  SCN4A  voltage-gated sodium channel activity  muscle contraction   
240  SCN4A  voltage-gated sodium channel activity  muscle contraction   
241  SCN4A  voltage-gated sodium channel activity  muscle contraction   
242  SCN4A  voltage-gated sodium channel activity  muscle cont

Again, we get back multiple paths with the same answer node (node_0) due to differences in edge sources.  However, we can see that there are two genes that share this set of GO terms, SCN4A and SCN7A.

We can also have more than one unspecified node.  Suppose, for instance that we wanted to do the previous query, but we also wanted to know what chemicals interact with the genes that we find.  We can do another star query, where one of our spokes is unspecified:

In [21]:
go_terms=['GO:0005248','GO:0006936','GO:0005244','GO:0034765','GO:0035725','GO:0019228','GO:0086010','GO:0006814',None]
types = ['biological_process_or_activity' for i in range(8)]+['chemical_substance']
star_q_compound = make_star_question(types,go_terms,'gene',rebuild=True)
star_q_compound

{'machine_question': {'edges': [{'source_id': 0, 'target_id': 1},
   {'source_id': 0, 'target_id': 2},
   {'source_id': 0, 'target_id': 3},
   {'source_id': 0, 'target_id': 4},
   {'source_id': 0, 'target_id': 5},
   {'source_id': 0, 'target_id': 6},
   {'source_id': 0, 'target_id': 7},
   {'source_id': 0, 'target_id': 8},
   {'source_id': 0, 'target_id': 9}],
  'nodes': [{'id': 0, 'type': 'gene'},
   {'curie': 'GO:0005248', 'id': 1, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006936', 'id': 2, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0005244', 'id': 3, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0034765', 'id': 4, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0035725', 'id': 5, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0019228', 'id': 6, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0086010', 'id': 7, 'type': 'biological_process_or_activity'},
   {'curie': 'GO:0006814', 'id': 8, 'type': '

**WARNING: This takes a long time to run**

In [22]:
common_gene_compound_answer = quick(star_q_compound)

Return Status: 200


In [23]:
cgc_nodes = extract_node_names(common_gene_compound_answer)
cgc_nodes[['node_0','node_9']].drop_duplicates()

node_0       node_9
0  SCN4A  lamotrigine

In [24]:
len(cgc_nodes)

250

Many other graph shapes are possible, these are simply a couple of examples.  For instance, we could add onto this query an edge from node 0 to a new disease node (node 10) and also include an edge from the target node (node 9) to this same disease node. Now we would be finding genes that share a set of GO terms, then looking for disease,drug pairs such that the disease, the drug, and the gene are all interconnected.

### Controlling the number of results

Even though we asked for 250 rows, we only recieved a single chemical back, due to the variations in the many edges in the graph.    To get more results, we can send a `max_results` parameter to quick.  Here we will send 20000.

In [25]:
common_gene_compound_answer_2k = quick(star_q_compound,max_results=20000)


Return Status: 200


In [26]:
cgc_nodes_2k = extract_node_names(common_gene_compound_answer_2k)
cgc_nodes_2k[['node_0','node_9']].drop_duplicates()

node_0                                             node_9
0      SCN4A                                        lamotrigine
256    SCN4A                                      Carbamazepine
768    SCN4A                                         benzocaine
1024   SCN4A                                          Lidocaine
1536   SCN4A                                          phenytoin
2560   SCN4A                                          Tocainide
2816   SCN4A                                         Mexiletine
3072   SCN7A                                         benzocaine
3328   SCN4A                                         Flecainide
3584   SCN4A                                         Zonisamide
3840   SCN7A                                             Nickel
4352   SCN7A                                          Tocainide
4608   SCN4A                                        mepivacaine
4864   SCN4A                                      Valproic Acid
5376   SCN4A                                       pilsicainide
5632   SCN4A                                         rufinamide
5888   SCN4A                                           3-cresol
6144   SCN4A                                         etidocaine
6400   SCN4A                                       orphenadrine
6656   SCN4A                                     pirinixic acid
7168   SCN4A                                       procainamide
7424   SCN4A                                       troglitazone
7552   SCN4A  N-(2-methyl-3-(4-(4-(4-(trifluoromethoxy)benzy...
7616   SCN4A                                     4-chlorophenol
7744   SCN4A                            2-methyl-4-chlorophenol
8960   SCN4A                                   (+)-JQ1 compound
9088   SCN4A                                        2,6-xylenol
9728   SCN4A                                      chloroxylenol
10240  SCN4A                                        ropivacaine
10304  SCN4A  3-[4-(4-Fluoro-phenoxy)-phenyl]-pyrazole-1-car...
10368  SCN4A                                     Zetekitoxin AB
10432  SCN4A                                                 NA
10624  SCN4A        3-[4-(4-Fluoro-phenoxy)-phenyl]-1H-pyrazole
10688  SCN4A  (E)-2-(4-(4-fluorophenoxy)benzylidene)hydrazin...
11264  SCN4A  (R)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
11456  SCN4A  (S)-2-(2,6-dimethylphenylthio)-1-phenylethanamine
11520  SCN4A                            eslicarbazepine acetate
11584  SCN4A                                           Ethotoin
11840  SCN4A  (R)-1-(2,6-dimethylphenoxy)-3-phenylpropan-2-a...
11904  SCN4A  (S)-3-(4-chloro-2,6-dimethylphenoxy)-2-methylp...
12096  SCN4A     (S)-2-(2,6-dimethylphenoxy)-1-phenylethanamine
12160  SCN4A          (S)-1-phenyl-3-(o-tolyloxy)propan-2-amine
12352  SCN4A  (R)-3-(4-chloro-2,6-dimethylphenoxy)-2-methylp...
12800  SCN4A  (S)-3-(2,6-dimethylphenylthio)-2-methylpropan-...
12928  SCN4A                                        Fospropofol
19840  SCN4A  3-[4-(4-Nitro-phenoxy)-phenyl]-pyrazole-1-carb...
19904  SCN4A  3-[4-(2,4-Difluoro-phenoxy)-phenyl]-pyrazole-1...
19968  SCN4A  3-(4-Phenoxy-phenyl)-pyrazole-1-carboxylic aci...

## Rebuild and caching

For details on ROBOKOP caching, and guidance for when to use the `rebuild` parameter, see the notebook on the expand service.  Briefly, the quick service will only run against the cache unless the rebuild parameter is passed, in which case services are re-queried.   The rebuild parameter is passed as part of the json query as seen in the `make_star_query` function above.  Another example is seen below in the Edge Properties section.

## Edge Properties

In the expand notebook, we saw how a particular predicate could be chosen to limit the results.  This same type information can in fact be added to any edge in a quick query, providing fine control to the results returned.